### ISSUE TITLE
Invalid URI: The Uri string is too long. (Xamarin.Forms)

### ISSUE THREAD

**Issue by gaf09** [reactiveui/refit#296]
Hi,

I need to send a request where it sends (among other data) an image in base64 blob. The problem is that the amount of characters in the image is somewhat large, but not much.
When I send more than 2,000 characters (approximately), the following error occurs when I start the request `Invalid URI: The Uri string is too long.`.

My interface code:
```
public interface RestServiceApi
{
	[Post("/webservices/rest.php?version=1.0")]
	Task<MyResponse> sendRequest([Body(BodySerializationMethod.UrlEncoded)] MyRequest request);
}
```

In the model MyRequest is where the image its found:
```
public class MyRequest
{
	public string blob { get; set; }
        public string mimetype { get; set; }
        // others
}
```

And the error occurs when sending the request, as I said ... if I send a very small image it is sent without any problem. The problem lies in having several characters to form the URI.
```
var response = await restService.sendRequest(request);
```

I appreciate your help !
Greetings,
Gabriel

**Comment by Cheesebaron** [comment]
Don't send stuff like that as base64 encoded url parameters, as you just experienced there is a limit to how long they can be.

**Comment by gaf09** [comment]
@Cheesebaron thanks for your answer !

I'm restricted. The Webservice is external and the only way that allows me is so in base64.
Is there no way to avoid this limitation? Note: Im relatively new with Xamarin and C#.

**Comment by Cheesebaron** [comment]
This is a limitation in accordance to [RFC2616](http://www.faqs.org/rfcs/rfc2616.html) to support all combinations of client and servers.

**Comment by onovotny** [comment]
Most browsers do support longer URI's FWIW, especially if not entered directly in the address bar -- i.e., 302 redirects, links, etc.

**Comment by onovotny** [comment]
I'm not sure where that limit you're seeing is coming from though; I don't think HttpClient has any such underlying limits.

**Comment by Cheesebaron** [comment]
@onovotny pretty sure it happens here: https://github.com/paulcbetts/refit/blob/master/Refit/RequestBuilderImplementation.cs#L217

Uri AFAIK is limited to... 2083 characters or something like that.

**Comment by Bartmax** [comment]
I'm facing the same problem and I don't understand why it's an URI length error when the data is supposed to be in the body. What am I missing?

I need to send a POST (form-encoded) body to a `/short/endpoint`.

Is it possible using refit? Isn't this a bug?

**Comment by Bartmax** [comment]
This is a bug here : https://github.com/dotnet/corefx/issues/1936

A minimal repro:
```
var sb = new StringBuilder();
for (int i = 0; i < 10000; i++)
{
   sb.Append("llllllllloooonnnnnggggsssssstttttrrrrrriiiiinnnnnggg");
}
var content = new FormUrlEncodedContent(new List<KeyValuePair<string, string>> { { new KeyValuePair<string, string>("name", sb.ToString()) } });
```

while this is not directly related to Refit, I wanted to post here for people that may land here and also in case Refit want to implement it's own type similar to `FormUrlEncodeContent`

**Comment by onovotny** [comment]
@Bartmax Certainly happy to take an alternative to `FormUrlEncodedContent`, would you be up for a PR to do this?

**Comment by Bartmax** [comment]
I definitely can try to put the custom one i'm using on my app into refit.

**Comment by Sylix** [comment]
I solve this problem by changing the parameter at the interface, from BodySerializationMethod.**UrlEncoded** to BodySerializationMethod.**Json**

`Task<MyResponse> sendRequest([Body(BodySerializationMethod.UrlEncoded)] MyRequest request);`

Changes to 

`Task<MyResponse> sendRequest([Body(BodySerializationMethod.Json)] MyRequest request);`

**Comment by DerekSMorin** [comment]
I was having trouble trying to debug against refit.

But I was able to work around this making a CustomFormUrlEncodedContent similar to what @Bartmax  had done.

I tweaked his a bit to use the answer from:

https://stackoverflow.com/questions/6695208/uri-escapedatastring-invalid-uri-the-uri-string-is-too-long

The idea is to split a large string into chunks that are small enough to call Uri.EscapeDataString on.

So for now I changed my signature to:

`Task<String> RunServerScript([AliasAs("tid")] string tenantID, [Body] CustomFormUrlEncodedContent data);
`
But I would have been able to use:

`Task<String> RunServerScript([AliasAs("tid")] string tenantID, Body(BodySerializationMethod.UrlEncoded)] Dictionary<string, object> data);`

If refit was changed to use CustomFormUrlEncodedContent internally.
